index.js

```diff
@@ -0,0 +1,734 @@
+  //. | [Chain][]       |     ✔︎     |           |           |           |
+  //. | [ChainRec][]    |     ✔︎     |           |           |           |
+  //. | [Monad][]       |     ✔︎     |           |           |           |
+  //. | [Foldable][]    |     ✔︎     |           |           |           |
+  //. | [Traversable][] |     ✔︎     |           |           |           |
+  //. | [Extend][]      |           |           |           |           |
+  //. | [Comonad][]     |           |           |           |           |
+
+  var implementations = {
+
+    //  Arguments#equals :: Arguments ~> Arguments -> Boolean
+    'Arguments#fantasy-land/equals': function(other) {
+      return other.length === this.length &&
+             every_.call(this,
+                         function(x, idx) { return equals(other[idx], x); });
+    },
+
+    //  Array#toString :: Array a ~> String
+    'Array#toString': function() {
+      return '[' + this.map(toString).join(', ') + ']';
+    },
+
+    //  Array#equals :: Array a ~> Array a -> Boolean
+    'Array#fantasy-land/equals': function(other) {
+      return other.length === this.length &&
+             this.every(function(x, idx) { return equals(other[idx], x); });
+    },
+
+    //  Array#concat :: Array a ~> Array a -> Array a
+    'Array#fantasy-land/concat': function(xs) { return this.concat(xs); },
+
+    //  Array#empty :: Array a ~> Array a
+    'Array#fantasy-land/empty': function() { return []; },
+
+    //  Array#map :: Array a ~> (a -> b) -> Array b
+    'Array#fantasy-land/map': function(f) {
+      return this.map(function(x) { return f(x); });
+    },
+
+    //  Array#ap :: Array a ~> Array (a -> b) -> Array b
+    'Array#fantasy-land/ap': function(fs) {
+      var xs = this;
+      return chain(function(f) { return map(f, xs); }, fs);
+    },
+
+    //  Array#of :: Array a ~> b -> Array b
+    //
+    //  This method is never actually used, but it must be defined in order
+    //  for `Applicative.test([])` to evaluate `true` in environments without
+    //  [`Array.of`][Array.of].
+    'Array#fantasy-land/of': /* istanbul ignore next */ function(x) {
+      return [x];
+    },
+
+    //  Array#chain :: Array a ~> (a -> Array b) -> Array b
+    'Array#fantasy-land/chain': function(f) {
+      var result = [];
+      this.forEach(function(x) { push_.apply(result, f(x)); });
+      return result;
+    },
+
+    //  Array.chainRec :: ((a -> c, b -> c, a) -> Array c, a) -> Array b
+    'Array.fantasy-land/chainRec': function(f, x) {
+      var next = function(x) { return {value: x, done: false}; };
+      var done = function(x) { return {value: x, done: true}; };
+      var $todo = [x];
+      var $done = [];
+      while ($todo.length > 0) {
+        var $buckets = [[], []];
+        f(next, done, $todo.shift()).forEach(function(x) {
+          $buckets[x.done ? 1 : 0].push(x.value);
+        });
+        unshift_.apply($todo, $buckets[0]);
+        push_.apply($done, $buckets[1]);
```

Comment:
I tried your suggestion initially, but `Array.prototype.unshift.apply(xs, [1, 2, 3])` and `xs.unshift(1); xs.unshift(2); xs.unshift(3)` are not equivalent (even ignoring mutation) because the elements are added backwards with the second approach.

I like the idea of using a `for` loop here. I'll push a commit shortly.